### PR TITLE
SONAR-7590 Redirect to requested page with identity provider

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

**Comment by julienlancelot** [comment]
@ehartmann Please review ! (I will add an IT)

File: server/sonar-server/src/main/java/org/sonar/server/authentication/InitFilter.java [hunk 1]
```diff
@@ -47,13 +47,15 @@
   private final BaseContextFactory baseContextFactory;
   private final OAuth2ContextFactory oAuth2ContextFactory;
   private final AuthenticationEvent authenticationEvent;
+  private final OAuth2Redirection oAuthRedirection;
 
   public InitFilter(IdentityProviderRepository identityProviderRepository, BaseContextFactory baseContextFactory,
-    OAuth2ContextFactory oAuth2ContextFactory, Server server, AuthenticationEvent authenticationEvent) {
+                    OAuth2ContextFactory oAuth2ContextFactory, Server server, AuthenticationEvent authenticationEvent, OAuth2Redirection oAuthRedirection) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/authentication/InitFilter.java]: bad tab :-)

Comment (julienlancelot) [reply server/sonar-server/src/main/java/org/sonar/server/authentication/InitFilter.java]: Fixed !

File: server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2ContextFactory.java [hunk 2]
```diff
@@ -19,9 +19,15 @@
  */
 package org.sonar.server.authentication;
 
+import static java.lang.String.format;
```
Comment (ehartmann) [inline server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2ContextFactory.java]: static import should be the last ones

File: server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2CallbackFilter.java [hunk 3]
```diff
@@ -19,7 +19,13 @@
  */
 package org.sonar.server.authentication;
 
+import static java.lang.String.format;
```
Comment (ehartmann) [inline server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2CallbackFilter.java]: static import should be the last ones

File: server/sonar-server/src/main/java/org/sonar/server/authentication/InitFilter.java [hunk 4]
```diff
@@ -19,14 +19,21 @@
  */
 package org.sonar.server.authentication;
 
+import static java.lang.String.format;
```
Comment (ehartmann) [inline server/sonar-server/src/main/java/org/sonar/server/authentication/InitFilter.java]: static import should be the last ones

File: server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2ContextFactory.java [hunk 5]
```diff
@@ -115,7 +119,9 @@ public void verifyCsrfState() {
     @Override
     public void redirectToRequestedPage() {
       try {
-        getResponse().sendRedirect(server.getContextPath() + "/");
+        String redirectTo = oAuthRedirection.getAndDelete(request, response);
+        String redirectToUrl = redirectTo == null ? "/" : redirectTo;
```
Comment (ehartmann) [inline server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2ContextFactory.java]: I think server.getContextPath() is missing when redirectTo is null, isn't it ?

File: server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2Redirection.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    response.addCookie(newCookieBuilder(request)
+      .setName(REDIRECT_TO_COOKIE)
+      .setValue(redirectTo)
+      .setHttpOnly(true)
+      .setExpiry(-1)
+      .build());
+  }
+
+  @CheckForNull
```
Comment (ehartmann) [inline server/sonar-server/src/main/java/org/sonar/server/authentication/OAuth2Redirection.java]: Wouldn't it better to have an Optional<String> ?

**Pull request merged**